code/datums/radiation_wave.dm

```diff
@@ -1,95 +1,232 @@
+	var/list/intensity_new[(distance + 1) * 8]
+	// "Class" it belongs to
+	var/branchclass = 2 ** round(log(2, distance))
+	// The secondary i, or the offset for i
+	var/j
 
-	var/strength
-	if(steps>1)
-		strength = INVERSE_SQUARE(intensity, max(range_modifier*steps, 1), 1)
-	else
-		strength = intensity
+	for(var/i in 1 to distance * 8)
+		//Culls invalid intensities
+		if(cintensity[i] * falloff < RAD_WAVE_MINIMUM)
+			continue
+		var/xpos
+		var/ypos
+		switch(i / distance)
+			if(0 to 2)
+				//Yes it starts one step off of what you'd expect. Blame BYOND.
+				xpos = cmaster_turf.x + distance
+				ypos = cmaster_turf.y + distance - i
+			if(2 to 4)
+				xpos = cmaster_turf.x + distance * 3 - i
+				ypos = cmaster_turf.y - distance
+			if(4 to 6)
+				xpos = cmaster_turf.x - distance
+				ypos = cmaster_turf.y - distance * 5 + i
+			if(6 to 8)
+				xpos = cmaster_turf.x - distance * 7 + i
+				ypos = cmaster_turf.y + distance
+		//Culls invalid coords
+		if(xpos < 1 || xpos > world.maxx || ypos < 1 || ypos > world.maxy)
+			continue
 
-	if(strength < RAD_WAVE_MINIMUM)
-		qdel(src)
-		return
+		//The radiation is considered alive
+		futile = FALSE
+		var/turf/place = locate(xpos, ypos, cmaster_turf.z)
+		atoms = get_rad_contents(place)
 
-	radiate(atoms, FLOOR(strength, 1))
+		//Actual radiation spending
+		cintensity[i] *= radiate(atoms, cintensity[i] * falloff)
 
-	check_obstructions(atoms) // reduce our overall strength if there are radiation insulators
+		//Obstruction handling
+		check_obstructions(atoms, i)
 
-/datum/radiation_wave/proc/get_rad_atoms()
-	var/list/atoms = list()
-	var/distance = steps
-	var/cmove_dir = move_dir
-	var/cmaster_turf = master_turf
+		/*
+		 * This is what I call pseudo-raycasting (PRC). Real raycasting would be ridiculously expensive,
+		 * So this is the solution I came up with. Don't try to understand it by seeing the code.
+		 * You have been warned. If you find yourself really having to touch this cursed code,
+		 * consider axing this away before contacting me via git-fu email digging.
+		 *
+		 * Therefore, I urge you not to hastily assume this code a culprit of your problem.
+		 * This code is responsible just for *keeping the rads going forward* more reasonably
+		 * in regard to obstruction and contamination cost. But, of course, if you are rewriting
+		 * (notwithstanding how questionable rewriting something major of a mature codebase like
+		 * every normal SS13 codebase is) the entire radiation code, then this code should be
+		 * considered for deletion.
+		 *
+		 * On a side note, this implementation isn't very ideal. So please remove this instead of
+		 * trying to improve it when its time has come. (i.e. another total overhaul)
+		 *
+		 * ~Xenomedes, Christmas 2020
+		 */
 
-	if(cmove_dir == NORTH || cmove_dir == SOUTH)
-		distance-- //otherwise corners overlap
+		// Handling eight fundamental (read: perfectly straight) branches
+		if((j = i / distance) == (j = round(j)))
```

Comment:
Huh. This works?

Reply:
How it works:
1. it calculates the value of `i / distance`
2. the value is saved to the variable `j`
3. which also is a left side of the `==` operator
4. the evaluation goes to the right side of the `==`operator; the `round(j)` is calculated
5. and then the result is again saved to the variable `j`
6. which also is a right side of the `==` operator
7. then `==` operator does the evaluation: it ultimately checks whether `i / distance` and `round(i / distance)` are the same, while the value of the latter is saved to `j`
8. the code inside the `if` is evaluated if they are the same, or code inside the `else` if otherwise
9. the value `j` is used afterwards on both cases